File: composeApp/src/androidMain/kotlin/app/dmvtestprep/Extensions.kt
package app.dmvtestprep

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.LinkAnnotation
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withLink
import androidx.compose.ui.text.withStyle
import app.dmvtestprep.core.ui.model.ColorModel
import app.dmvtestprep.core.ui.model.ThemeColor
import app.dmvtestprep.domain.model.Question
import app.dmvtestprep.datasource.api.ApiConfig
import app.dmvtestprep.domain.model.MarkdownText

val Question.lsLabel: String
    get() = num.formatted(learningStage.type.value)

val Question.lsTextColor: Color
    @Composable
    get() = learningStage.getTextColor().asColor()

val Question.lsBackgroundColor: Color
    @Composable
    get() = learningStage.getBackgroundColor().asColor()

val Question.backgroundColor: Color
    @Composable
    get() = learningStage.getBackgroundColor().asColor(alpha = 0.1f)

val Question.imageUrl: String?
    get() = image?.toFullImageUrl()

@Composable
fun ThemeColor.asColor(alpha: Float? = null): Color {
    val isDark = isSystemInDarkTheme()
    val colorModel = if (isDark) dark else light

    return colorModel.asColor(alpha ?: this@asColor.alpha)
}

fun ColorModel.asColor(alpha: Float? = null): Color {
    return Color(r, g, b, alpha ?: a)
}

@Composable
fun AnnotatedString.Builder.AppendMarkdownTexts(texts: List<MarkdownText>) {
    texts.forEach {
        when (it) {
            is MarkdownText.Bold -> withStyle(SpanStyle(fontWeight = FontWeight.Medium)) {
                append(it.text)
            }
            is MarkdownText.Italic -> withStyle(SpanStyle(fontStyle = FontStyle.Italic)) {
                append(it.text)
            }
            is MarkdownText.Link -> withLink(LinkAnnotation.Url(url = it.url)) {
                withStyle(SpanStyle(textDecoration = TextDecoration.Underline, color = it.color.asColor())) {
                    append(it.text)
                }
            }
            is MarkdownText.Plain -> append(it.text)
        }
    }
}

fun String?.toFullImageUrl(): String? {
    return ApiConfig.getFullImageUrl(path = this)
}

File: iosApp/iosApp/Shared/Extensions.swift
import UIKit
import SwiftUICore
import Shared

extension Question {
    var lsLabel: String {
        num.formatted(suffix: learningStage.type.value)
    }
    var lsTextColor: Color {
        learningStage.getTextColor().asColor()
    }
    var lsBackgroundColor: Color {
        learningStage.getBackgroundColor().asColor()
    }
    var backgroundColor: Color {
        learningStage.getBackgroundColor().asColor(alpha: 0.1)
    }
    var imageUrl: String? {
        image.toFullImageUrl()
    }
}

extension ThemeColor {
    func asColor(alpha: Float? = nil) -> Color {
        Color(uiColor: UIColor { trait in
            let colorModel = (trait.userInterfaceStyle == .dark ? self.dark : self.light)
            
            return UIColor(
                red: CGFloat(colorModel.r),
                green: CGFloat(colorModel.g),
                blue: CGFloat(colorModel.b),
                alpha: CGFloat(alpha ?? self.alpha)
            )
        })
    }
}

extension [MarkdownText] {
    func toAttributedString() -> AttributedString {
        var attributedString = AttributedString()

        for text in self {
            var attrText = AttributedString(text.text)

            switch text {
            case is MarkdownText.Bold:
                attrText.font = .boldSystemFont(ofSize: 16)
            case is MarkdownText.Italic:
                attrText.font = .italicSystemFont(ofSize: 16)
            case let link as MarkdownText.Link:
                attrText.font = .system(size: 16)
                attrText.foregroundColor = link.color.asColor()
                attrText.underlineStyle = Text.LineStyle.single
                attrText.link = URL(string: link.url)
            default:
                break
            }

            attributedString.append(attrText)
        }
        
        return attributedString
    }
}

extension Int {
    func toInt32() -> Int32 {
        return Int32(self)
    }
}

extension Int32 {
    func toInt() -> Int {
        return Int(self)
    }
    func toKotlinInt() -> KotlinInt {
        return KotlinInt(int: self)
    }
}

extension String? {
    func toFullImageUrl() -> String? {
        return ApiConfig.shared.getFullImageUrl(path: self)
    }
}

public extension View {
    func transparentScrolling() -> some View {
        modifier(TransparentScrollingModifier())
    }
}

private struct TransparentScrollingModifier: ViewModifier {
    func body(content: Content) -> some View {
        if #available(iOS 16.0, *) {
            content.scrollContentBackground(.hidden)
        } else {
            content.onAppear {
                UITextView.appearance().backgroundColor = .clear
            }
            .onDisappear {
                UITextView.appearance().backgroundColor = nil
            }
        }
    }
}